### DISCUSSION TITLE
Pydantic import error in fastapi

### DISCUSSION THREAD

**Discussion by Han-aorweb** [pydantic/pydantic#5185] [Projects using Pydantic]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

I am using Heroku web hosting, this problem only occurs when I try to run the `heroku local` which is running my API application on my local environment, it never appears in other cases. I am also using poetry to manage the packages, at first, I thought it was a poetry problem, but when I used a brand new venv and install all packages manually, the problem was still unfixed. And there is a similar issue but for me, it is that my apple Mac is using arm64, but in the error, it demands x86_64, which is the weird part.  

```
(env) wangweihan@Jays-MacBook-Air fast_api % heroku local                                                       
4:44:35 PM web.1 |  Traceback (most recent call last):
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/bin/uvicorn", line 8, in <module>
4:44:35 PM web.1 |      sys.exit(main())
4:44:35 PM web.1 |               ^^^^^^
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
4:44:35 PM web.1 |      return self.main(*args, **kwargs)
4:44:35 PM web.1 |             ^^^^^^^^^^^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/click/core.py", line 1055, in main
4:44:35 PM web.1 |      rv = self.invoke(ctx)
4:44:35 PM web.1 |           ^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
4:44:35 PM web.1 |      return ctx.invoke(self.callback, **ctx.params)
4:44:35 PM web.1 |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/click/core.py", line 760, in invoke
4:44:35 PM web.1 |      return __callback(*args, **kwargs)
4:44:35 PM web.1 |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/uvicorn/main.py", line 404, in main
4:44:35 PM web.1 |      run(
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/uvicorn/main.py", line 569, in run
4:44:35 PM web.1 |      server.run()
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/uvicorn/server.py", line 60, in run
4:44:35 PM web.1 |      return asyncio.run(self.serve(sockets=sockets))
4:44:35 PM web.1 |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 190, in run
4:44:35 PM web.1 |      return runner.run(main)
4:44:35 PM web.1 |             ^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 118, in run
4:44:35 PM web.1 |      return self._loop.run_until_complete(task)
4:44:35 PM web.1 |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/uvicorn/server.py", line 67, in serve
4:44:35 PM web.1 |      config.load()
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/uvicorn/config.py", line 477, in load
4:44:35 PM web.1 |      self.loaded_app = import_from_string(self.app)
4:44:35 PM web.1 |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/uvicorn/importer.py", line 24, in import_from_string
4:44:35 PM web.1 |      raise exc from None
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/uvicorn/importer.py", line 21, in import_from_string
4:44:35 PM web.1 |      module = importlib.import_module(module_str)
4:44:35 PM web.1 |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/importlib/__init__.py", line 126, in import_module
4:44:35 PM web.1 |      return _bootstrap._gcd_import(name[level:], package, level)
4:44:35 PM web.1 |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
4:44:35 PM web.1 |    File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
4:44:35 PM web.1 |    File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
4:44:35 PM web.1 |    File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
4:44:35 PM web.1 |    File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
4:44:35 PM web.1 |    File "<frozen importlib._bootstrap_external>", line 940, in exec_module
4:44:35 PM web.1 |    File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/app/main.py", line 4, in <module>
4:44:35 PM web.1 |      from fastapi import Depends, FastAPI, HTTPException
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/fastapi/__init__.py", line 7, in <module>
4:44:35 PM web.1 |      from .applications import FastAPI as FastAPI
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/fastapi/applications.py", line 15, in <module>
4:44:35 PM web.1 |      from fastapi import routing
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/fastapi/routing.py", line 22, in <module>
4:44:35 PM web.1 |      from fastapi import params
4:44:35 PM web.1 |    File "/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/fastapi/params.py", line 4, in <module>
4:44:35 PM web.1 |      from pydantic.fields import FieldInfo, Undefined
4:44:35 PM web.1 |  ImportError: dlopen(/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/pydantic/__init__.cpython-311-darwin.so, 0x0002): tried: '/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/pydantic/__init__.cpython-311-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/wangweihan/Desktop/Straford_Project/Stratford_u
4:44:35 PM web.1 |  >  rbansim/fast_api/env/lib/python3.11/site-packages/pydantic/__init__.cpython-311-darwin.so' (no such file), '/Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/pydantic/__init__.cpython-311-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64'))
[DONE] Killing all processes with signal  SIGINT
4:44:35 PM web.1 Exited with exit code null
```

### Example Code

```Python
from pydantic.fields import FieldInfo, Undefined
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.6
            pydantic compiled: True
                 install path: /Users/wangweihan/Desktop/Straford_Project/Stratford_urbansim/fast_api/env/lib/python3.11/site-packages/pydantic
               python version: 3.11.0 (v3.11.0:deaf509e8f, Oct 24 2022, 14:43:23) [Clang 13.0.0 (clang-1300.0.29.30)]
                     platform: macOS-13.2.1-arm64-arm-64bit
     optional deps. installed: ['dotenv', 'email-validator', 'typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [X] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by Han-aorweb** [comment]
I have tried uninstall, reinstall pydantic, and install the speicfic "arm64" wheel, not working.

**Reply by seanbetts** [reply]
I've tried this as well and still not able to get it working

**Comment by yezz123** [comment]
This error message typically indicates that you are trying to use pydantic compiled for a different architecture than the one your Python interpreter is currently running on.

> This can happen, for example, if you are trying to run Python code on an Apple M1 Mac, which uses the 'arm64' architecture, but the library or module you are using was compiled for the older 'x86_64' architecture that was used on older Macs.

you can try to create a Dockerfile that specifies the architecture and dependencies that you need. 

Here's an example Dockerfile that builds an image for Python 3.11 running on the 'arm64' architecture:

```dockerfile
FROM python:3.11-alpine

RUN apk add --no-cache build-base

COPY requirements.txt /app/requirements.txt

RUN pip install --no-cache-dir -r /app/requirements.txt

WORKDIR /app

CMD ["python", "app.py"]
```

In my case, I updated both my packages and some projects in PROD, but it had no impact. Docker can primarily verify the events that occurred in the vicinity.

**Comment by dmontagu** [comment]
I am removing the bug label as this is not a pydantic bug, but is an issue with the docker setup and/or whatever is happening with heroku. I'm also going to convert this to a discussion as it does not appear to be specific to pydantic.

I would encourage you to look into how the base image that you are using was built (and possibly adapt it to run on arm64), and/or look into the `platform` argument to docker. Basically you just need to figure out where the architecture mis-match is coming from. (It might also have something to do with how `heroku local` works, which I don't know much about.)

**Reply by seanbetts** [reply]
Hi, I'm having the same issue on my local machine (M1 mac), so not sure it is an architecture issue?

**Reply by dmontagu** [reply]
I think this will be hard for anyone to help debug unless you can share a repository where you can reproduce the issue, and provide instructions for how to reproduce it (assuming using an M1 mac).

If you can do that, I suspect it should be pretty straightforward to resolve (again, relates to needing to specify the platform properly at some point in the build process).

**Reply by seanbetts** [reply]
Thanks for getting back to me, makes sense. I am on an M1 mac and running in a virtual environment and have tried multiple versions of python using pyenv, so it must be to do with how some packages run on arm64. The issue seems to be with __init__.cpython-311-darwin.so and needing the arm64 version but I've tried forcing the arm64 version of pydantic and still getting the same error:

`ImportError: dlopen(.../venv/lib/python3.11/site-packages/pydantic/__init__.cpython-311-darwin.so, 0x0002): tried: '.../venv/lib/python3.11/site-packages/pydantic/__init__.cpython-311-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64')), '.../venv/lib/python3.11/site-packages/pydantic/__init__.cpython-311-darwin.so' (no such file), '.../venv/lib/python3.11/site-packages/pydantic/__init__.cpython-311-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64'))`

**Comment by seanbetts** [comment]
I am having the exact same issue as this whilst running it on my local machine. @Han-aorweb - did you find a solution?

**Reply by Han-aorweb** [reply]
I changed my app to docker based and decided to change from Heroku to other services, so I don't need to debug this problem anymore, but I do think is something wrong with architecture compatibility with some packages, I m not sure which package specifically causing this problem, what were you intending to do before encountering this error?

**Reply by seanbetts** [reply]
I'm just building a Vue + FastAPI app and using Supertokens for authentication. This issue originally came up as I was having issues with uvicorn but as I've been trying to work out a fix I've started getting pydantic errors. I'm running in a virtual environment using pyenv and tried multiple versions of python, so I'm guessing it must be an issue with some packages not playing well with arm64 as I'm on an M1 mac.

**Comment by AngleZhou** [comment]
There is a python3.12-intel64 binary with 1 architectur x86_64, under /Library/Frameworks/Python.framework/Versions/3.12/bin(python installation path). 
I change the interpreter to this binary, then the error is gone.

**Reply by wcohnshb** [reply]
Hi Angle, how did you change the interpreter for the binary?
